In [1]:
%load_ext autoreload
%autoreload 2

In [33]:
import os

import numpy as np
import pandas as pd

from tqdm import tqdm
from joblib import Parallel, delayed

In [3]:
from utils import load_config, perform_statistical_tests
from data.load.data_loader import get_codecfake_audio_id_list, load_parquet_data
from features import (
    LowLevelFeatureExtractor, 
    HighLevelFeatureExtractor, 
    plot_low_level_feature_dist, 
    plot_high_level_feature_dist, 
    perform_pca_and_plot
)

In [4]:
config    = load_config()
cache_dir = config['data_paths']['codecfake']['cache_files']
features_dir = config['data_paths']['features']
audio_ids = get_codecfake_audio_id_list()

#### Codecfake - Partitions: 0, 1, ... 379

In [17]:
285 + 95

380

ajay: np.arange(0, 95) --> 0, 1, ..., 94

keerthana: np.arange(95, 190) --> 95, 96, ..., 189

Ruohe: np.arange(190, 285) --> 190, 191, ..., 284

Prudhvi: np.arange(285, 380) --> 285, 286, ..., 379

In [19]:
partitions = np.arange(0, 95)
partitions

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94])

In [20]:
def generate_dataframe(iterable_ds):
    rows = []
    for audio in iterable_ds:
        audio_id  = audio['audio_id']
        audio_arr = audio['audio']['array']
        srate     = audio['audio']['sampling_rate']
        real_fake = audio['real_or_fake']
    
        rows.append({
            'audio_id': audio_id,
            'audio_arr': audio_arr,
            'srate': srate,
            'real_or_fake': real_fake
        })
    return df

In [24]:
df_list = []

for partition_id in tqdm(partitions, total=len(partitions), desc="Processing partitions"):
    iterable_ds = load_parquet_data(partition_id=partition_id)    
    partition_df = generate_dataframe(iterable_ds)
    df_list.append(partition_df)

df = pd.concat(df_list, ignore_index=True)

df

Processing partitions: 100%|██████████| 95/95 [07:13<00:00,  4.56s/it]


audio_id                                          audio_arr  srate  \
0      p225_002  [-0.0045166015625, -0.00665283203125, -0.00607...  48000   
1      p225_002  [0.001953125, 0.001556396484375, 0.00164794921...  16000   
2      p225_002  [-0.001220703125, -0.001129150390625, -0.00103...  24000   
3      p225_002  [0.001861572265625, 0.001922607421875, 0.00195...  16000   
4      p225_002  [-0.00335693359375, -0.0032958984375, -0.00320...  48000   
...         ...                                                ...    ...   
61840  p225_191  [0.00762939453125, 0.007568359375, 0.007446289...  48000   
61841  p225_191  [0.00787353515625, 0.00799560546875, 0.0081176...  16000   
61842  p225_191  [0.002777099609375, 0.001678466796875, 0.00183...  24000   
61843  p225_191  [0.00311279296875, 0.003204345703125, 0.002960...  16000   
61844  p225_191  [0.00439453125, 0.0057373046875, 0.00509643554...  48000   

      real_or_fake  
0                R  
1              F03  
2              F04  
3              F01  
4              F05  
...            ...  
61840          F05  
61841          F01  
61842          F04  
61843          F03  
61844            R  

[61845 rows x 4 columns]

#### Extract Features

In [27]:
audio_processor = LowLevelFeatureExtractor(target_sr=16000, include_only=['spectral', 'prosodic', 'voice_quality'])
feature_computer = HighLevelFeatureExtractor()

In [28]:
low_level_gen        = audio_processor.low_level_feature_generator(df.iloc[:20])
high_level_features  = list(feature_computer.high_level_feature_generator(low_level_gen))
high_level_features_df = pd.DataFrame(high_level_features)
high_level_features_df

Processing Audios: 100%|██████████| 20/20 [00:23<00:00,  1.19s/it]


audio_id real_or_fake  spectral_centroid_mean  spectral_centroid_std  \
0   p225_002            R             1407.277479            1139.511158   
1   p225_002          F03             1269.150323            1085.214354   
2   p225_002          F04             1336.049769            1171.331560   
3   p225_002          F01             1540.345739            1212.507952   
4   p225_002          F05             1414.717840            1149.669533   
5   p225_002          F02             1514.202295            1179.175846   
6   p225_002          F06             1269.150323            1085.214354   
7   p225_003            R             1724.882050            1358.306823   
8   p225_003          F04             1686.359745            1392.656929   
9   p225_003          F03             1700.782291            1308.502013   
10  p225_003          F06             1700.782291            1308.502013   
11  p225_003          F02             1791.209249            1408.466867   
12  p225_003          F05             1740.279417            1358.008199   
13  p225_003          F01             1867.926207            1478.520182   
14  p225_005          F04             1553.217764            1131.104607   
15  p225_005          F03             1581.314247            1099.550237   
16  p225_005            R             1605.708773            1099.336588   
17  p225_005          F06             1581.314247            1099.550237   
18  p225_005          F02             1677.229296            1156.060546   
19  p225_005          F05             1608.073380            1074.416395   

    spectral_centroid_var  spectral_centroid_min  spectral_centroid_max  \
0            1.298486e+06             273.979621            4860.736866   
1            1.177690e+06             319.092868            4752.988182   
2            1.372018e+06             247.256673            4922.406802   
3            1.470176e+06             257.468837            5074.111834   
4            1.321740e+06             295.103707            4982.683428   
5            1.390456e+06             275.770102            5091.169775   
6            1.177690e+06             319.092868            4752.988182   
7            1.844997e+06             240.330124            5445.572581   
8            1.939493e+06             168.940520            5577.680477   
9            1.712178e+06             289.027907            5388.677656   
10           1.712178e+06             289.027907            5388.677656   
11           1.983779e+06             236.626324            5761.676749   
12           1.844186e+06             277.044296            5580.742488   
13           2.186022e+06             257.929241            5768.435948   
14           1.279398e+06             209.275100            5329.144045   
15           1.209011e+06             453.713344            5294.671488   
16           1.208541e+06             301.962245            5259.902010   
17           1.209011e+06             453.713344            5294.671488   
18           1.336476e+06             274.291738            5681.442518   
19           1.154371e+06             313.896633            5236.565437   

    spectral_centroid_range  spectral_centroid_25th_percentile  \
0               4586.757245                         538.847218   
1               4433.895314                         501.144619   
2               4675.150129                         433.114366   
3               4816.642997                         688.193002   
4               4687.579722                         604.047516   
5               4815.399673                         634.001211   
6               4433.895314                         501.144619   
7               5205.242457                         856.539372   
8               5408.739957                         814.810942   
9               5099.649749                         845.365038   
10              5099.649749                         845.365038   
11              5525.050425            

### Using Parallel Processing

In [34]:
def extract_features(row, audio_processor, feature_computer):
    low_level_features = audio_processor.extract_features(row)
    high_level_features = feature_computer.compute_high_level_features(low_level_features)
    return high_level_features

audio_processor  = LowLevelFeatureExtractor(target_sr=16000, include_only=['spectral', 'prosodic', 'voice_quality'])
feature_computer = HighLevelFeatureExtractor()

high_level_features   = Parallel(n_jobs=5)(
    delayed(extract_features)(row, audio_processor, feature_computer) 
    for _, row in df.iloc[:1000].iterrows()
)
high_level_feature_df = pd.DataFrame(high_level_features)
high_level_feature_df.to_csv(os.path.join(features_dir, 'sample_features_1000.csv'), index=False)